tests/testsuite/build_dir.rs

```diff
@@ -0,0 +1,589 @@
+        .file(
+            ".cargo/config.toml",
+            r#"
+            [build]
+            target-dir = "build"
+            "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert_build_dir_layout(p.root().join("build"), "debug");
+    assert_exists_patterns_with_base_dir(
+        &p.root(),
+        &[
+            &format!("build/debug/deps/foo*{EXE_SUFFIX}"),
+            "build/debug/deps/foo*.d",
+            "build/debug/.fingerprint/foo-*/bin-foo",
+            "build/debug/.fingerprint/foo-*/bin-foo.json",
+            "build/debug/.fingerprint/foo-*/dep-bin-foo",
+            "build/debug/.fingerprint/foo-*/invoked.timestamp",
+        ],
+    );
+    assert_not_exists(&p.root().join("target"));
+}
+
+#[cargo_test]
+fn libs() {
+    // https://doc.rust-lang.org/reference/linkage.html#r-link.staticlib
+    let (staticlib_prefix, staticlib_suffix) =
+        if cfg!(target_os = "windows") && cfg!(target_env = "msvc") {
+            ("", ".lib")
+        } else {
+            ("lib", ".a")
+        };
+
+    // (crate-type, list of final artifacts)
+    let lib_types = [
+        ("lib", ["libfoo.rlib", "libfoo.d"]),
+        (
+            "dylib",
+            [
+                &format!("{DLL_PREFIX}foo{DLL_SUFFIX}"),
+                &format!("{DLL_PREFIX}foo.d"),
+            ],
+        ),
+        (
+            "cdylib",
+            [
+                &format!("{DLL_PREFIX}foo{DLL_SUFFIX}"),
+                &format!("{DLL_PREFIX}foo.d"),
+            ],
+        ),
+        (
+            "staticlib",
+            [
+                &format!("{staticlib_prefix}foo{staticlib_suffix}"),
+                &format!("{staticlib_prefix}foo.d"),
+            ],
+        ),
+    ];
+
+    for (lib_type, expected_files) in lib_types {
+        let p = project()
+            .file("src/lib.rs", r#"fn foo() { println!("Hello, World!") }"#)
+            .file(
+                "Cargo.toml",
+                &format!(
+                    r#"
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+            edition = "2015"
+
+            [lib]
+            crate-type = ["{lib_type}"]
+            "#
+                ),
+            )
+            .file(
+                ".cargo/config.toml",
+                r#"
+            [build]
+            target-dir = "build"
+            "#,
+            )
+            .build();
+
+        p.cargo("build")
+            .masquerade_as_nightly_cargo(&[])
+            .enable_mac_dsym()
+            .run();
+
+        assert_build_dir_layout(p.root().join("build"), "debug");
+        assert_exists_patterns_with_base_dir(&p.root().join("build/debug"), &expected_files);
+        assert_not_exists(&p.root().join("target"));
+    }
+}
+
+#[cargo_test]
+fn binary_with_release() {
```

Comment:
I feel like this should be next to `binary_with_debug` as they seem to be closely related